### PR TITLE
Implement fetching server capabilities before the first server call is made

### LINKED ISSUE #1312: Call getSystemInfo before a first client call to obtain server capabilities
**Is your feature request related to a problem? Please describe.**
Temporal Server API provides new `getSystemInfo` call that returns a set of features supported by the server. 

**Describe the solution you'd like**
This call needs to be performed by the stubs or client before any other client call.

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Adds a new interceptor that fetches systemCapabilities before the first client call can be made.

Closes #1312

### CODE REVIEW

File: temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java [hunk 1]
```diff
@@ -140,7 +145,9 @@ private Channel applyHeadStandardInterceptors(Channel channel) {
     headers.put(CLIENT_NAME_HEADER_KEY, CLIENT_NAME_HEADER_VALUE);
 
     return ClientInterceptors.intercept(
-        channel, MetadataUtils.newAttachHeadersInterceptor(headers));
+        channel,
+        MetadataUtils.newAttachHeadersInterceptor(headers),
+        new SystemInfoInterceptor(systemCapabilitiesFuture));
```
Comment (Sushisource) [inline temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java:150]: Just double checking that these two will go first before any user-specified interceptors

**Review by Sushisource — APPROVED** [review]

File: temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java (hunk 1 above)
Comment (Spikhalskiy) [reply temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java:150]: yup

File: temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                    if (Status.UNIMPLEMENTED.getCode().equals(status.getCode())) {
+                      serverCapabilitiesFuture.complete(
+                          GetSystemInfoResponse.Capabilities.getDefaultInstance());
+                    }
+                    super.onClose(status, trailers);
+                  }
+                };
+          } else {
+            // Need to reach system capabilities, so make a getSystemInfo call in a blocking manner.
+            // We don't try to squash into one and optimize the several getSystemInfo calls that may
```
Comment (cretz) [inline temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java:74]: > We don't try to squash into one and optimize the several getSystemInfo calls

:+1: Same as Go. If in a racy situation the this code is reached by multiple calls and we have to make multiple getSystemInfo calls and overwrite that future one right after another, that's much more preferable than locking. https://github.com/temporalio/sdk-go/blob/942edfb5dce0dbff49e19f002470726660d54b53/internal/internal_workflow_client.go#L901-L904

**Review by cretz — APPROVED** [review]

File: temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java (hunk 2 above)
Comment (Spikhalskiy) [reply temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java:74]: Yeah, in Java the first who completes wins. Great, thanks for confirming that it's aligned!

**Pull request merged**